### PR TITLE
bpo-34755: Add few minor optimizations in _asynciomodule.c.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue34755

### CODE REVIEW

File: Modules/_asynciomodule.c [hunk 1]
```diff
@@ -2702,163 +2690,138 @@ task_step_impl(TaskObj *task, PyObject *exc)
...
-        goto fail;
-    }
-    res = PyObject_IsTrue(o);
-    Py_DECREF(o);
-    if (res == -1) {
-        /* An exception while checking if 'val' is True */
-        goto fail;
-    }
-    if (res == 1) {
+    if (PyGen_Check(result)) {
```
Comment (1st1) [inline Modules/_asynciomodule.c]: This is incorrect; the old code called `inspect.isgenerator` on purpose as it recognizes generator-like objects compiled with Cython.  I suggest to first call `PyGen_Check` (fast path) and then `inspect.isgenerator`.

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Modules/_asynciomodule.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_asynciomodule.c]: `inspect.isgenerator(obj)` is literally `isinstance(obj, types.GeneratorType)`. This is translated to C as `PyObject_IsInstance(obj, &PyGen_Type)`. This is almost equivalent to `PyObject_TypeCheck(obj, &PyGen_Type)`, but falls back to `issubclass(obj.__class__, types.GeneratorType)` if `PyObject_TypeCheck` returns false. `PyGen_Check(obj)` is expanded to just `PyObject_TypeCheck(obj, &PyGen_Type)`.

Thus there is a difference if the generator-like object itself is not an instance of a GeneratorType subclass, but sets a fake `__class__`.

Comment (1st1) [reply Modules/_asynciomodule.c]: Ah, sorry Serhiy, you're right.  I worked before on multiple issues with Cython and I think we fixed them by adding `isinstance(o, collections.abc.Generator)` or `isinstance(o, collections.abc.Coroutine)`.  Anyways, never mind, please go ahead with this PR.

**Review by 1st1 — APPROVED** [review]

File: Modules/_asynciomodule.c (hunk 1 above)
Comment (serhiy-storchaka) [reply Modules/_asynciomodule.c]: Should we use `PyGen_Check()` or `PyObject_IsInstance()`? The only difference is that different error messages are produced for generator-like objects with fake `__class__` (is this a case of Cython?). The code with `PyObject_IsInstance()` is more verbose.

Comment (1st1) [reply Modules/_asynciomodule.c]: We should try to maintain the exact same behaviour between Python and C versions so I'd say let's use `PyObject_IsInstance()` (even though this code path is for legacy coroutines and will eventually be removed).

**Comment by serhiy-storchaka** [comment]
Thank you for your review @1st1!

**Pull request merged**